### PR TITLE
gh-109408: Run `patchcheck` in GitHub Actions

### PR DESCRIPTION
**Pull request by sobolevn**
Continuing your idea in https://github.com/python/cpython/issues/109408#issuecomment-1720902187

Refs https://github.com/python/cpython/pull/109452

* Issue: gh-109408

### CODE REVIEW

File: .github/workflows/lint.yml [hunk 1]
```diff
@@ -24,3 +24,22 @@ jobs:
...
+        with:
+          fetch-depth: 0
+          ref: ${{ github.event.pull_request.head.ref }}
+          repository: ${{ github.event.pull_request.head.repo.full_name }}
+      - uses: actions/setup-python@v4
+        with:
+          python-version: "3.11"
+      - name: "Run patchcheck"
+        run: |
+          git status && git branch
```
Comment (vstinner) [inline .github/workflows/lint.yml]: Why not putting them on two lines?

File: .github/workflows/lint.yml [hunk 2]
```diff
@@ -24,3 +24,22 @@ jobs:
...
+  patchcheck:
+    if: "github.repository == 'python/cpython' && github.event_name == 'pull_request'"
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
+          ref: ${{ github.event.pull_request.head.ref }}
```
Comment (vstinner) [inline .github/workflows/lint.yml]: The job fails: need a git checkout to get modified files
Getting base branch for PR ... not a PR branch

Comment (sobolevn) [reply .github/workflows/lint.yml]: I finally found what's wrong:

```
 Getting base branch for PR ... No /opt/hostedtoolcache/Python/3.11.5/x64/lib/python3.11/config-3.11-x86_64-linux-gnu
not a PR branch
```

`SRCDIR = sysconfig.get_config_var('srcdir')` does not have `.git` directory, that's why everything fails :)

File: .github/workflows/lint.yml [hunk 3]
```diff
@@ -24,3 +24,22 @@ jobs:
...
+
+  patchcheck:
+    if: "github.repository == 'python/cpython' && github.event_name == 'pull_request'"
+    runs-on: ubuntu-latest
+    timeout-minutes: 10
+
+    steps:
+      - uses: actions/checkout@v4
+        with:
+          fetch-depth: 0
```
Comment (hugovk) [inline .github/workflows/lint.yml]: Can we avoid checking out the full 33 years and 120k commit history?

CAM originally had the same in https://github.com/python/cpython/pull/108065 but figured out a way around it.

File: .github/workflows/lint.yml [hunk 4]
```diff
@@ -24,3 +24,22 @@ jobs:
...
+          ref: ${{ github.event.pull_request.head.ref }}
+          repository: ${{ github.event.pull_request.head.repo.full_name }}
```
Comment (AA-Turner) [inline .github/workflows/lint.yml]: Why is this needed? We're always checking out CPython, right?

File: .github/workflows/lint.yml (hunk 1 above)
Comment (AA-Turner) [reply .github/workflows/lint.yml]: Should we make this a distinct step for "git info"? I'm unsure what it achieves, though.

File: Tools/patchcheck/patchcheck.py
Comment (AA-Turner) [inline Tools/patchcheck/patchcheck.py:1]: I assume the added prints are just temporary for debugging

**Comment by sobolevn** [comment]
Finally! :)

File: .github/workflows/build.yml [hunk 5]
```diff
@@ -143,6 +143,11 @@ jobs:
           make regen-deepfreeze
           make -j4 regen-all
           make regen-stdlib-module-names
+      - name: Run patchcheck
+        if: github.event_name == 'pull_request'
+        run: |
+          git fetch origin
```
Comment (hugovk) [inline .github/workflows/build.yml:149]: Do we need to fetch origin? It takes 1m 49s for this step.

We don't do it on [Azure Pipelines and `patchcheck` takes 2s](https://dev.azure.com/Python/cpython/_build/results?buildId=137844&view=logs&j=256d7e09-002a-52d7-8661-29ee3960640e&t=3d7276d3-4e8d-5309-55ad-fb0b172d9925).

**Comment by AA-Turner** [comment]
@sobolevn -- if you want to keep ``patchcheck`` in its own job for speed, https://github.com/AA-Turner/cpython/commit/8af0f2c431ca213756c3bbd96882e7ee9bf200cf is a sketch of an approach.

Two changes are needed to ``patchcheck``, both due to prior assumptions that ``patchcheck`` runs on a local build of CPython.

- On CI, use the current working directory as ``SRCDIR`` rather than using ``sysconfig``.
- We use ``patchlevel.h`` to get the Python version, not ``sys.version_info`` (the docs use this approach too)

It may be worth considering using the ``CI`` environment variable rather than a ``ci=true`` CLI flag, but that could be delayed to a later date.

A

File: .github/workflows/build.yml (hunk 5 above)
Comment (sobolevn) [reply .github/workflows/build.yml:149]: Let's try :)

I think that we might need it because of the heavy `git` machinery inside `patchcheck`.
I think we might need it during backports for older branches.

**Comment by sobolevn** [comment]
@AA-Turner I am not quite comfortable refactoring code that I don't quite understand :)
And since this is liteally the first time I see `patchcheck`, I would prefer to keep this PR as simple as possible.

Later we can change the tooling to be more convenient / faster, but for now I would like to stick with my solution.

File: .github/workflows/build.yml (hunk 5 above)
Comment (sobolevn) [reply .github/workflows/build.yml:149]: Nope, it does not work:

```
Run # git fetch origin
Checked 107 modules (31 built-in, 75 shared, 1 n/a on linux-x86_64, 0 disabled, 0 missing, 0 failed on import)
LD_LIBRARY_PATH=/home/runner/work/cpython/cpython ./python ./Tools/patchcheck/patchcheck.py --ci true
Getting base branch for PR ... origin/main
fatal: ambiguous argument 'origin/main': unknown revision or path not in the working tree.
Use '--' to separate paths from revisions, like this:
'git <command> [<revision>...] -- [<file>...]'
error running git diff --name-status origin/main
make: *** [Makefile:2914: patchcheck] Error 1
Getting the list of files that have been added/changed ... 
```

**Comment by hugovk** [comment]
In the end I'd like to remove or replace as much of patchcheck with equivalents that we do understand, like pre-commit or Ruff linting. But this is a good first step.

**Comment by sobolevn** [comment]
I agree 100%, that what I was thinking about all the way :)

**Comment by hugovk** [comment]
> @sobolevn -- if you want to keep `patchcheck` in its own job for speed, [AA-Turner@8af0f2c](https://github.com/AA-Turner/cpython/commit/8af0f2c431ca213756c3bbd96882e7ee9bf200cf) is a sketch of an approach.

This approach takes [22s](https://github.com/AA-Turner/cpython/actions/runs/6204967853/job/16847629844?pr=13), including fetching source and installing a prebuilt Python.

This PR piggy backs on 'Check if generated files are up to date' to take advantage of a local Python build, but is [+1m 48s](https://github.com/python/cpython/actions/runs/6206082751/job/16849941429?pr=109459), mostly fetching source.

> * We use `patchlevel.h` to get the Python version, not `sys.version_info` (the docs use this approach too)

Could be nice to refactor `Doc/tools/extensions/patchlevel.py` and put it somewhere common for use by both?

**Comment by sobolevn** [comment]
@hugovk I propose we merge this as-is for now, because `+1m43s` is not even close to how long our regular tests take (around `15m 4s`). Moreover, we will save `+13m` from Azure Pipeline, when duplication is deleted:
Снимок экрана 2023-09-16 в 15 02 35


Later we can:
- Replace most checks with `pre-commit` / `ruff`
- Refactor the rest checks to use modern alternatives / API
- Do not build CPython from source here for no reason

**Comment by AA-Turner** [comment]
A slightly different approach would be https://github.com/AA-Turner/cpython/commit/5ad4fafd83b0f3bc70e5f9db109899e0c5a615e4, which avoids all of the Git chicanery and just runs patchcheck on all files in the source tree. This seems to take ~20s ([ex1](https://github.com/AA-Turner/cpython/actions/runs/6212403134/job/16862494208?pr=14), [ex2](https://github.com/AA-Turner/cpython/actions/runs/6212403134/job/16862494208?pr=14)).

A

**Comment by hugovk** [comment]
> A slightly different approach would be [AA-Turner@5ad4faf](https://github.com/AA-Turner/cpython/commit/5ad4fafd83b0f3bc70e5f9db109899e0c5a615e4), which avoids all of the Git chicanery and just runs patchcheck on all files in the source tree. This seems to take ~20s ([ex1](https://github.com/AA-Turner/cpython/actions/runs/6212403134/job/16862494208?pr=14), [ex2](https://github.com/AA-Turner/cpython/actions/runs/6212403134/job/16862494208?pr=14)).
> 
> A

~30s for the whole thing, but that's not too bad and I think I prefer this approach -- we also run on all files for pre-commit on the CI.

**Comment by sobolevn** [comment]
Yeap, all files is also fine by me (since they are all correct anyway).

I will rework @AA-Turner's script to make it possible to run on Windows as well, because it might be useful for local testing. Thanks!

**Comment by AA-Turner** [comment]
> I will rework @AA-Turner's script to make it possible to run on Windows as well

I developed it on Windows--though I normalised all paths to forward-slash, for simplicity.

A

**Comment by AA-Turner** [comment]
If all three of #109854; #109890; and #109891 are merged then this PR may be closed.

A
